// express-openapi-validator/dist/middlewares/openapi.security.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
const defaultSecurityHandler = (req, scopes, schema) => true;
function security(context, securityHandlers) {
    return async (req, res, next) => {
        var _a, _b, _c, _d;
        // TODO move the folllowing 3 check conditions to a dedicated upstream middleware
        if (!req.openapi) {
            // this path was not found in open api and
            // this path is not defined under an openapi base path
            // skip it
            return next();
        }
        const openapi = req.openapi;
        const expressRoute = openapi.expressRoute;
        if (!expressRoute) {
            return next(util_1.validationError(404, req.path, 'not found'));
        }
        const pathSchema = openapi.schema;
        if (!pathSchema) {
            // add openapi metadata to make this case more clear
            // its not obvious that missig schema means methodNotAllowed
            return next(util_1.validationError(405, req.path, `${req.method} method not allowed`));
        }
        // use the local security object or fallbac to api doc's security or undefined
        const securities = (_a = openapi.schema.security, (_a !== null && _a !== void 0 ? _a : context.apiDoc.security));
        const path = openapi.openApiRoute;
        if (!path || !Array.isArray(securities) || securities.length === 0) {
            return next();
        }
        const securitySchemes = (_b = context.apiDoc.components) === null || _b === void 0 ? void 0 : _b.securitySchemes;
        if (!securitySchemes) {
            const message = `security referenced at path ${path}, but not defined in 'components.securitySchemes'`;
            return next(util_1.validationError(500, path, message));
        }
        try {
            const results = await new SecuritySchemes(securitySchemes, securityHandlers, securities).executeHandlers(req);
            // TODO handle AND'd and OR'd security
            // This assumes OR only! i.e. at least one security passed authentication
            let firstError = null;
            let success = false;
            for (const r of results) {
                if (r.success) {
                    success = true;
                    break;
                }
                else if (!firstError) {
                    firstError = r;
                }
            }
            if (success)
                next();
            else
                throw firstError;
        }
        catch (e) {
            const message = ((_d = (_c = e) === null || _c === void 0 ? void 0 : _c.error) === null || _d === void 0 ? void 0 : _d.message) || 'unauthorized';
            const err = util_1.validationError(e.status, path, message);
            next(err);
        }
    };
}
exports.security = security;
class SecuritySchemes {
    constructor(securitySchemes, securityHandlers, securities) {
        this.securitySchemes = securitySchemes;
        this.securityHandlers = securityHandlers;
        this.securities = securities;
    }
    async executeHandlers(req) {
        // use a fallback handler if security handlers is not specified
        // This means if security handlers is specified, the user must define
        // all security handlers
        const fallbackHandler = !this.securityHandlers
            ? defaultSecurityHandler
            : null;
        const promises = this.securities.map(async (s) => {
            var _a, _b, _c;
            try {
                if (Util.isEmptyObject(s)) {
                    // anonumous security
                    return { success: true };
                }
                const securityKey = Object.keys(s)[0];
                const scheme = this.securitySchemes[securityKey];
                const handler = (_b = (_a = this.securityHandlers) === null || _a === void 0 ? void 0 : _a[securityKey], (_b !== null && _b !== void 0 ? _b : fallbackHandler));
                const scopesTmp = s[securityKey];
                const scopes = Array.isArray(scopesTmp) ? scopesTmp : [];
                if (!scheme) {
                    const message = `components.securitySchemes.${securityKey} does not exist`;
                    throw { status: 500, message };
                }
                if (!scheme.hasOwnProperty('type')) {
                    const message = `components.securitySchemes.${securityKey} must have property 'type'`;
                    throw { status: 500, message };
                }
                if (!handler) {
                    const message = `a security handler for '${securityKey}' does not exist`;
                    throw { status: 500, message };
                }
                new AuthValidator(req, scheme, scopes).validate();
                // expected handler results are:
                // - throw exception,
                // - return true,
                // - return Promise<true>,
                // - return false,
                // - return Promise<false>
                // everything else should be treated as false
                const securityScheme = scheme;
                const success = await handler(req, scopes, securityScheme);
                if (success === true) {
                    return { success };
                }
                else {
                    throw Error();
                }
            }
            catch (e) {
                return {
                    success: false,
                    status: (_c = e.status, (_c !== null && _c !== void 0 ? _c : 401)),
                    error: e,
                };
            }
        });
        return Promise.all(promises);
    }
}
class AuthValidator {
    constructor(req, scheme, scopes = []) {
        const openapi = req.openapi;
        this.req = req;
        this.scheme = scheme;
        this.path = openapi.openApiRoute;
        this.scopes = scopes;
    }
    validate() {
        this.validateApiKey();
        this.validateHttp();
        this.validateOauth2();
        this.validateOpenID();
    }
    validateOauth2() {
        const { req, scheme, path } = this;
        if (['oauth2'].includes(scheme.type.toLowerCase())) {
            // TODO oauth2 validation
        }
    }
    validateOpenID() {
        const { req, scheme, path } = this;
        if (['openIdConnect'].includes(scheme.type.toLowerCase())) {
            // TODO openidconnect validation
        }
    }
    validateHttp() {
        const { req, scheme, path } = this;
        if (['http'].includes(scheme.type.toLowerCase())) {
            const authHeader = req.headers['authorization'] &&
                req.headers['authorization'].toLowerCase();
            if (!authHeader) {
                throw Error(`Authorization header required`);
            }
            const type = scheme.scheme && scheme.scheme.toLowerCase();
            if (type === 'bearer' && !authHeader.includes('bearer')) {
                throw Error(`Authorization header with scheme 'Bearer' required`);
            }
            if (type === 'basic' && !authHeader.includes('basic')) {
                throw Error(`Authorization header with scheme 'Basic' required`);
            }
            this.dissallowScopes();
        }
    }
    validateApiKey() {
        const { req, scheme, path } = this;
        if (scheme.type === 'apiKey') {
            if (scheme.in === 'header') {
                if (!req.headers[scheme.name.toLowerCase()]) {
                    throw Error(`'${scheme.name}' header required`);
                }
            }
            else if (scheme.in === 'query') {
                if (!req.query[scheme.name]) {
                    throw Error(`query parameter '${scheme.name}' required`);
                }
            }
            // TODO scheme in cookie
            this.dissallowScopes();
        }
    }
    dissallowScopes() {
        if (this.scopes.length > 0) {
            // https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#security-requirement-object
            throw {
                status: 500,
                message: "scopes array must be empty for security type 'http'",
            };
        }
    }
}
class Util {
    static isEmptyObject(o) {
        return (typeof o === 'object' &&
            Object.entries(o).length === 0 &&
            o.constructor === Object);
    }
}
